### PR TITLE
Include enforcement date in warnings for future effective_on violations

### PR DESCRIPTION
**Pull request by Acepresso**
When the CLI demotes a deny result to a warning due to a future effective_on date, the warning message is now enhanced to indicate when enforcement begins.

Before: "The required 'cpe' label is missing"
After:  "The required 'cpe' label is missing. This will become a failure starting on 2025-01-01T00:00:00Z"

This applies to both sources of effective_on: rule annotations and rule_data items. The effective_on date is included as-is (RFC3339 format), matching the pattern used by trusted_task's future_deny_rule warnings.
No duplicate warnings for trusted_task rules. The enhancement only applies to failures being demoted, not to existing warning rules like `future_deny_rule`.

  ## Test plan
  
  - [ ] Unit tests for `formatFutureEnforcementMessage` covering: valid effective_on, missing metadata, non-string value, empty string, nil metadata
  - [ ] Existing `TestConftestEvaluatorEvaluateSeverity` updated and passing
  - [ ] Full evaluator unit test suite passes with no regressions
  - [ ] Run `ec validate` against a policy with a future `effective_on` date and verify the warning message includes the enforcement date

Ref: https://redhat.atlassian.net/browse/EC-1901
Assisted-by: Claude Opus 4.6

### CODE REVIEW

File: internal/evaluator/filters.go [hunk 1]
```diff
@@ -1204,3 +1218,21 @@ func (f *UnifiedPostEvaluationFilter) determineOriginalType(filteredResult Resul
...
+// Returns the enhanced message if effective_on is present, otherwise returns
+// the original message unchanged.
+func formatFutureEnforcementMessage(result Result) string {
+	originalMessage := result.Message
+
+	// Check if this result has an effective_on date
+	effectiveOnStr, ok := result.Metadata[metadataEffectiveOn].(string)
+	if !ok || effectiveOnStr == "" {
+		return originalMessage
+	}
```
[the remark this answers is not in the document]
Comment (Acepresso) [reply internal/evaluator/filters.go]: Fixed.

File: internal/evaluator/filters.go [hunk 2]
```diff
@@ -997,7 +997,14 @@ func (f *LegacyPostEvaluationFilter) CategorizeResults(
 			}
 		case "failure":
 			if getSeverity(result) == severityWarning || !isResultEffective(result, effectiveTime) {
-				warnings = append(warnings, result)
+				// If being demoted due to future effective_on, enhance the message
```
[the remark this answers is not in the document]
Comment (Acepresso) [reply internal/evaluator/filters.go]: legit. I fixed it.

File: internal/evaluator/filters.go [hunk 3]
```diff
@@ -1204,3 +1210,21 @@ func (f *UnifiedPostEvaluationFilter) determineOriginalType(filteredResult Resul
```
[the remark this answers is not in the document]
Comment (Acepresso) [reply internal/evaluator/filters.go:1217]: fixed

File: internal/evaluator/filters.go [hunk 4]
```diff
@@ -1204,3 +1210,21 @@ func (f *UnifiedPostEvaluationFilter) determineOriginalType(filteredResult Resul
 
 	return "unknown"
 }
+
```
[the remark this answers is not in the document]
Comment (Acepresso) [reply internal/evaluator/filters.go]: fixed

File: internal/evaluator/filters.go [hunk 5]
```diff
@@ -996,8 +996,13 @@ func (f *LegacyPostEvaluationFilter) CategorizeResults(
 				warnings = append(warnings, result)
 			}
 		case "failure":
-			if getSeverity(result) == severityWarning || !isResultEffective(result, effectiveTime) {
+			if getSeverity(result) == severityWarning {
 				warnings = append(warnings, result)
+			} else if !isResultEffective(result, effectiveTime) {
```
[the remark this answers is not in the document]
Comment (Acepresso) [reply internal/evaluator/filters.go:1008]: that comment is obsolete, already fixed

**Review by robnester-rh — COMMENTED** [review]
~Moved to inline comment~

File: internal/evaluator/filters.go [hunk 6]
```diff
@@ -996,7 +1003,10 @@ func (f *LegacyPostEvaluationFilter) CategorizeResults(
 				warnings = append(warnings, result)
 			}
 		case "failure":
-			if getSeverity(result) == severityWarning || !isResultEffective(result, effectiveTime) {
+			if getSeverity(result) == severityWarning {
```
Comment (robnester-rh) [inline internal/evaluator/filters.go:1006]: Nit: the unchecked type assertion `result.Metadata[metadataEffectiveOn].(string)` could panic if the value is unexpectedly non-string or missing — a comma-ok form would be more defensive, though the control flow via `isResultEffective` makes this safe in practice. Same applies to the identical assertion in the Unified filter below.

**Review by robnester-rh — DISMISSED** [review]
LGTM

File: internal/evaluator/filters.go (hunk 6 above)
Comment (Acepresso) [reply internal/evaluator/filters.go:1006]: initially I did it as you suggested inside `formatEffectiveOnMessage()`, but then realized that it is already checked by `isResultEffective()`.
Perhaps I can add a test case that covers this edge case, and leave the code as is, wdyt?

Comment (robnester-rh) [reply internal/evaluator/filters.go:1006]: I think that would be sufficient.

Comment (Acepresso) [reply internal/evaluator/filters.go:1006]: Thanks, I added a new commit with this test case.

File: internal/evaluator/filters.go [hunk 7]
```diff
@@ -996,7 +1003,10 @@ func (f *LegacyPostEvaluationFilter) CategorizeResults(
...
+				warnings = append(warnings, result)
+			} else if !isResultEffective(result, effectiveTime) {
+				result.Message = formatEffectiveOnMessage(result.Message, result.Metadata[metadataEffectiveOn].(string))
```
[the remark this answers is not in the document]
Comment (Acepresso) [reply internal/evaluator/filters.go:1009]: this is covered by a unit-test

File: internal/evaluator/filters_test.go [hunk 8]
```diff
@@ -1796,3 +1980,36 @@ func TestPipelineIntentionWithMultipleValues(t *testing.T) {
 			"security package should be included (has included rules)")
 	})
 }
+
+//////////////////////////////////////////////////////////////////////////////
+// formatEffectiveOnMessage tests
+//////////////////////////////////////////////////////////////////////////////
+
+func TestFormatFutureEnforcementMessage(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (Acepresso) [reply internal/evaluator/filters_test.go]: Fixed.

File: internal/evaluator/filters.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (Acepresso) [reply internal/evaluator/filters.go:1006]: there is a test case that covers this edge case.

File: internal/evaluator/filters_test.go [hunk 9]
```diff
@@ -758,6 +758,190 @@ func TestUnifiedPostEvaluationFilter(t *testing.T) {
...
+			if tc.expectWarning {
+				assert.Len(t, warnings, 1)
+				assert.Empty(t, failures)
+				if tc.expectEnhanced {
+					assert.Equal(t, tc.expectedMessage, warnings[0].Message)
+				}
+			} else {
+				assert.Empty(t, warnings)
+				assert.Len(t, failures, 1)
+			}
```
[the remark this answers is not in the document]
Comment (Acepresso) [reply internal/evaluator/filters_test.go:940]: this is not true, the effectiveOn field is set in line 903.

File: internal/evaluator/filters.go (hunk 6 above)
Comment (Acepresso) [reply internal/evaluator/filters.go:1006]: @robnester-rh may I resolve this thread?

File: internal/evaluator/filters_test.go [hunk 10]
```diff
@@ -758,6 +758,190 @@ func TestUnifiedPostEvaluationFilter(t *testing.T) {
...
+		expectWarning   bool
+		expectEnhanced  bool
+		expectedMessage string
+	}{
+		{
+			name:          "severity override demotes failure to warning",
+			effectiveTime: now,
+			effectiveOn:   futureDate,
+			severity:      "warning",
+			expectWarning: true,
```
Comment (st3penta) [inline internal/evaluator/filters_test.go:790]: Consider adding the expectedMessage check here, to ensure it doesn't append the `This will become ...` string

**Review by st3penta — APPROVED** [review]
LGTM, just one nitpick in tests

File: internal/evaluator/filters_test.go (hunk 10 above)
Comment (Acepresso) [reply internal/evaluator/filters_test.go:790]: already covered - this is controlled by the flag `expectEnhanced` (which is `false` by default):
```
            if tc.expectWarning {
                assert.Len(t, warnings, 1)
                assert.Empty(t, failures)
                if tc.expectEnhanced {
                    assert.Equal(t, tc.expectedMessage, warnings[0].Message)                                                                                                          
                }
```
